### ISSUE TITLE
Crash due to unresolved dereference in cast to union

### ISSUE THREAD

**Issue by Bo-Yuan-Huang** [esbmc/esbmc#709]
Version: current master (`1f5b4ed3fbdb51a9280911f411b4766c4963a097`)

Command: 
``` bash
esbmc main.c --compact-trace
```
Input program:

``` c
#include <assert.h>

typedef struct {
  int data;
} s1;

typedef union {
  struct {
    int
      p0 : 1,
      p1 : 31;
  };
  int raw;
} s2;

int main() {
  s1 x;
  if (((s2)(&x)->data).p1) {}

  int var, ta, tb;
  ta = var;
  var = 0;
  tb = var;
  assert(ta == tb);

  return 0;
}
```

Output:
```
### ERROR:
Couldn't convert expression in unrecognised format
dereference
* pointer : address_of
    * pointer_obj : symbol
...
```

Note: The `--compact-trace` is used specifically for this small example. There's another larger example that also causes this similar issue without using the compact trace option.

**Comment by Bo-Yuan-Huang** [comment]
Sorry for keep throwing issues one after another, and thanks for the prompt responses on addressing the previous ones. We are adopting ESBMC to verify a massive codebase, and I won't know what comes next until I try with the fixes ...

**Comment by fbrausse** [comment]
`--compact-trace` implies `--no-slice`, the latter of which alone already exhibits the same behaviour as described in this issue since the no-op `if` statement in that case is not sliced out. The error occurs in smt_convt::convert_ast() where there should not be any dereference expression anymore. This slightly modified example has the same issue without `--no-slice` or `--compact-trace`:

```
#include <assert.h>

typedef struct {
  int data;
} s1;

typedef union {
  struct {
    int
      p0 : 1,
      p1 : 31;
  };
  int raw;
} s2;

int main() {
  s1 x;

  int var = ((s2)(&x)->data).p1;
  int ta, tb;
  ta = var;
  var = 0;
  tb = var;
  assert(ta == tb);

  return 0;
}
```

**Comment by fbrausse** [comment]
It is actually unrelated to bitfields or unions and looks more like an issue with compound literals. Same behaviour:
```c
#include <assert.h>

typedef struct {
  int data;
} s1;

typedef struct {
  int raw;
} s2;

int main() {
  s1 x;

  int var = (s2){(&x)->data}.raw;
  int ta, tb;
  ta = var;
  var = 0;
  tb = var;
  assert(ta == tb);

  return 0;
}
```

**Comment by lucasccordeiro** [comment]
> Sorry for keep throwing issues one after another, and thanks for the prompt responses on addressing the previous ones. We are adopting ESBMC to verify a massive codebase, and I won't know what comes next until I try with the fixes ...

@Bo-Yuan-Huang: you're more than welcome to open issues and interact with our team ;-) Are you verifying some open-source projects? We would appreciate hearing more about your experience using ESBMC.

**Comment by lucasccordeiro** [comment]
Some time ago, @fbrausse created an issue about compound literals: https://github.com/esbmc/esbmc/issues/591

**Comment by fbrausse** [comment]
@Bo-Yuan-Huang If you are in a position to modify the code, one option to work around this issue might be to replace all compound literals `(type){expression...}` as well as the compiler extension `(type)expression` for union types by first declaring and initializing a new variable `type new_name = {expression...}` and then using `new_name` in place of the compound literal.

**Comment by Bo-Yuan-Huang** [comment]
> @Bo-Yuan-Huang: you're more than welcome to open issues and interact with our team ;-) Are you verifying some open-source projects? We would appreciate hearing more about your experience using ESBMC.

Unfortunately, it's not an open-source project (yet) and I can only share the code patterns I distilled from the proprietary codebase. Would love to share my experiences and looking forward to any potential open-sourcing/collab in the future.

> @Bo-Yuan-Huang If you are in a position to modify the code, one option to work around this issue might be to replace all compound literals (type){expression...} as well as the compiler extension (type)expression for union types by first declaring and initializing a new variable type new_name = {expression...} and then using new_name in place of the compound literal.

Due to the code size and the frequent usage of compound literals, manual code modification may not be a practical option. It'd be very helpful to know if there's an option for Clang/LLVM to rewrite/avoid such a pattern.

**Comment by fbrausse** [comment]
After further investigation, this issue technically is about "[cast to union](https://gcc.gnu.org/onlinedocs/gcc/Cast-to-Union.html)" which is related to compound literals only in that atm. we handle both the [same](https://github.com/esbmc/esbmc/blob/1f5b4ed3fbdb51a9280911f411b4766c4963a097/src/clang-c-frontend/clang_c_convert.cpp#L1438) [way](https://github.com/esbmc/esbmc/blob/1f5b4ed3fbdb51a9280911f411b4766c4963a097/src/clang-c-frontend/clang_c_convert.cpp#L2219) (by converting them to constant_* expressions). Since @Bo-Yuan-Huang mentioned there are lots of cases of compound literals as well, I suppose we can keep the title.

**Comment by fbrausse** [comment]
I believe https://github.com/esbmc/esbmc/issues/591#issuecomment-1131638172 applies here:
> The error is coming from the smt backend, as it doesn't understand dereferences.
> I guess there are two possible solutions:
> 1. Teach the frontend to change the ast so the rest of esbmc understands it.
> 2. Look at why the pointer analysis/dereference code is not properly rewriting the expression.

Option 1. could be done by generating a new `lvalue = false` symbol for each "cast to union" in the frontend instead of building a constant_union expression. What is the expected impact on performance here? Do these non-`lvalue` symbols get encoded as variables for SMT?

I looked into option 2 and it appears that `dereference_expr_nonscalar()` is not prepared to work with unions (despite the `assert()` on the "nothing needs to be done" path claiming that would be the case for this constant_union2t expression). Could it be that `is_constant_expr()` is actually not meant to evaluate to true for aggregate types whose sub-expressions aren't constant as well?

**Comment by lucasccordeiro** [comment]
@fbrausse: I would suggest option 2 since you don't need to create a new expression, which would affect our CEX generation. @jmorse is our main contributor to the `dereference` module. I'll try to get @jmorse to answer your question.

@mikhailramalho: do you have any thoughts here?

**Comment by mikhailramalho** [comment]
I agree with @lucasccordeiro, the dereference code should be extended to understand constant_union2t expressions.

**Comment by jmorse** [comment]
Overall, everything is going to be simpler the earlier that dereferences get compiled out of the intermediate expressions. There are two awkward things happening here:
 * "constant" expressions really mean "literals", they don't necessarily have unchanging values,
 * In C++ at least, you can feed as complicated an expression into a literal as you like.
 * 
I suspect the dereference code needs to end up descending into constant_union2t's as they're really literals. However you then quickly get into the game of re-interpreting typed SMT symbols with arbitrary other types -- I think I skipped doing that for unions in the original implementation, I don't know what happens nowadays.

I guess my summary is, "yes union fields should be considered, but it's going to be ridiculously complicated".

**Comment by lucasccordeiro** [comment]
@Bo-Yuan-Huang: can I ask you whether you run the branch of https://github.com/esbmc/esbmc/pull/713 over your codebase? @fbrausse has proposed a fix for the issue reported here.

**Comment by Bo-Yuan-Huang** [comment]
> @Bo-Yuan-Huang: can I ask you whether you run the branch of #713 over your codebase? @fbrausse has proposed a fix for the issue reported here.

Thanks, I tried the fix and it solved the previous issue. However, it brought up a new issue that took me some time to distill a short example code. Since this issue was closed, I've created a new one (#719).
